socs/agents/lakeshore240/agent.py

```diff
@@ -1,82 +1,122 @@
 if not on_rtd:
-    from ocs import ocs_agent, site_config
-    from ocs.ocs_twisted import TimeoutLock
+    log = txaio.make_logger()  # pylint: disable=E1101
 
 
-class LS240_Agent:
+ActionReturnType = Optional[Dict[str, Any]]
+OcsOpReturnType = Tuple[bool, str]
+OcsInlineCallbackReturnType = Generator[Any, Any, OcsOpReturnType]
 
-    def __init__(self, agent, port="/dev/ttyUSB0", f_sample=2.5):
 
-        self.agent: ocs_agent.OCSAgent = agent
-        self.log = agent.log
-        self.lock = TimeoutLock()
+class Actions:
+    "Namespace to hold action classes for the Lakeshore240 agent."
 
-        self.port = port
-        self.module: Optional[Module] = None
+    class BaseAction:
+        "Base class for all actions."
 
-        self.f_sample = f_sample
+        def __post_init__(self):
+            self.deferred = defer.Deferred()
 
-        self.initialized = False
-        self.take_data = False
-
-        # Registers Temperature and Voltage feeds
-        agg_params = {
-            'frame_length': 60,
-        }
-        self.agent.register_feed('temperatures',
-                                 record=True,
-                                 agg_params=agg_params,
-                                 buffer_time=1)
-
-    # Task functions.
-    def init_lakeshore(self, session, params=None):
-        """init_lakeshore(auto_acquire=False)
-
-        **Task** - Perform first time setup of the Lakeshore 240 Module.
-
-        Parameters:
-            auto_acquire (bool, optional): Starts data acquisition after
-                initialization if True. Defaults to False.
+        def process(self, module: Module) -> ActionReturnType:
+            raise NotImplementedError
 
+    @dataclass
+    class UploadCalCurve(BaseAction):
+        """
+        Action to upload a calibration curve to the lakeshore.  For parameter
+        details, see the docstrings for the `upload_cal_curve` task.
         """
-        if params is None:
-            params = {}
 
-        auto_acquire = params.get('auto_acquire', False)
+        channel: int
+        filename: str
 
-        if self.initialized:
-            return True, "Already Initialized Module"
+        def process(self, module: Module) -> ActionReturnType:
+            log.info(f"Starting upload to channel {self.channel}...")
+            channel = module.channels[self.channel - 1]
+            channel.load_curve(self.filename)
+            time.sleep(0.1)
+            return None
 
-        with self.lock.acquire_timeout(0, job='init') as acquired:
-            if not acquired:
-                self.log.warn("Could not start init because "
-                              "{} is already running".format(self.lock.job))
-                return False, "Could not acquire lock."
+    @dataclass
+    class SetValues(BaseAction):
+        """
+        Action to configure settings on the lakeshore.  For parameter details,
+        see the docstrings for the `set_values` task.
+        """
```

Comment:
> Maybe I'm missing something, but I don't really see how the actions can be functions, since they have to actually store the data that tells the `main` process what parameters to use when processing the action.

Instead of storing the params as class attributes and putting the class in the queue you could put both the function and params in the queue, then use them in `_process_actions` -- kind of like how [`AgentOp` launches functions](https://github.com/simonsobs/ocs/blob/fe0342163e8a2480545df6015eddcfc86b97079b/ocs/ocs_agent.py#L982-L984). An example when I was thinking about this before the `Deferred` was dropped:
```python
def _process_actions(self, module: Module) -> None:
    """
    Processes queued actions using the provided Lakeshore Module.
    """
    while not self.action_queue.empty():
        # ActionClass(**params) -> (function, params, deferred)
        task, params, d_ = self.action_queue.get()
        try:
            log.info(f"Running action {task}")
            res = task(module, **params)
            d_.callback(res)
        except Exception as e:  # pylint: disable=broad-except
            log.error(f"Error processing action: {task}")
            d_.errback(e)
    return None
```

> The way I imagined solving this duplication of code is eliminating the task definition entirely (i.e. the `set_values` function in the LS240_Agent class), since this is almost entirely identical across different actions and is mostly boilerplate. Instead, we could have a `register_task_from_action` function, that took in the Action class and generated the OCS endpoint from it. Then we'd only have to document the parameters once, in the class definition.

Ah, that wasn't clear from the PR. This still seems like it would present as a class docstring, when we want to present a method docstring to users, since on the client end `process()` is hidden from them.

Maybe my above suggestion isn't the best either, but the launching machinery will be hidden from developers if this gets moved into a `register_task_from_action` function. I'm mostly concerned with the developer/user experience on the agent end.